aioshelly/wsrpc.py

```diff
@@ -0,0 +1,194 @@
+        await websocket.close()
+
+        for call_item in self._calls.items():
+            call_item.future.cancel()
+
+        self._calls = {}
+        self._rx_task = None
+
+    async def _handle_call(self, frame_id):
+        await self._websocket.send_json(
+            {
+                "id": frame_id,
+                "src": self._src,
+                "error": {"code": 500, "message": "Not Implemented"},
+            }
+        )
+
+    async def _handle_frame(self, frame):
+        if peer_src := frame.get("src", None):
+            if self._dst is not None and peer_src != self._dst:
+                _LOGGER.warning("Remote src changed: %s -> %s", self._dst, peer_src)
+            self._dst = peer_src
+
+        frame_id = frame.get("id", None)
+        method = frame.get("method", None)
+
+        if method:
+            # peer is invoking a method
+            params = frame.get("params", None)
+            if frame_id:
+                # and expects a response
+                _LOGGER.debug("handle call for frame_id: %s", frame_id)
+                await self._handle_call(frame_id)
+            else:
+                # this is a notification
+                _LOGGER.debug("Notification: %s %s", method, params)
+                await self._on_notification(method, params)
+
+        elif frame_id:
+            # looks like a response
+            if frame_id not in self._calls:
+                _LOGGER.warning("Response for an unknown request id: %s", frame_id)
+                return
+
+            call = self._calls.pop(frame_id)
+            call.resolve.set_result(frame)
+
+        else:
+            _LOGGER.warning("Invalid frame: %s", frame)
+
+    async def _rx_msgs(self):
+        async for msg in self._websocket:
+            _LOGGER.debug("Received Message: Type: %s,  Data: %s", msg.type, msg.data)
+
+            try:
+                frame = msg.json()
+            except ValueError as err:
+                raise InvalidMessage("Received invalid JSON.") from err
+            else:
+                await self._handle_frame(frame)
+
+        error = str(self._websocket.exception()) if self._websocket else "Disconnected"
+        _LOGGER.debug("Websocket error: %s", error)
+        await self._on_notification("WebSocketClosed", {"error": error})
+
+    @property
+    def connected(self) -> bool:
+        """Return if we're currently connected."""
+        return self._websocket is not None and not self._websocket.closed
+
+    async def call(self, method, params=None, timeout=10):
+        """Websocket RPC call."""
+        call = RPCCall(self._next_id, method, params)
+        call.resolve = asyncio.Future()
+        call.src = self._src
+        call.dst = self._dst
+
+        self._calls[call.call_id] = call
+        await self._websocket.send_json(call.request_frame)
+        call.sent_at = datetime.utcnow()
+
+        try:
+            resp = await asyncio.wait_for(call.resolve, timeout)
```

Comment:
We generally use `async_timeout`, it's a nicer syntax and is already installed as part of aiohttp;

```python
import async_timeout

async with async_timeout.timeout(timeout):
    await call.resolve
```